README.md
# Basil.js

The missing Javascript smart persistence layer.
Unified localstorage, cookie and session storage JavaScript API.


## Philosophy

Basil aims to ease the frontend storage management for developers. It strives
to be bulletproof and handle disabled cookies, full localStorage and other
unwanted native storage exceptions..

When you try to store something, basil will automatically look through all
the available storage mechanisms and find the best suited one to store your value. It
also handles storage of complex javascript objects using json.


## Basic Usage

```javascript
basil = new window.Basil(options);

// basic methods
basil.set('foo', 'bar'); // store 'bar' value under 'foo' key
basil.set('abc', 'xyz'); // store 'xyz' value under 'abc' key
basil.get('foo'); // returns 'bar'
basil.keys(); // returns ['abc', 'foo']
basil.keysMap(); // returns { 'abc': ['local'], 'foo': ['local'] }
basil.remove('foo'); // remove 'foo' value

// advanced methods
basil.check('local'); // boolean. Test if localStorage is available
basil.reset(); // reset all stored values under namespace
```

## Advanced Usage

### Storages

```javascript
basil = new window.Basil(options);

// force storage on the go through basil
// set 'bar' value under 'foo' key in localStorage
basil.set('foo', 'bar', { 'storages': ['local'] });

// set 'bar' value under 'foo' key.
// try first to store it into cookies and if not possible into localStorage
basil.set('foo', 'quux', { 'storages': ['cookie', 'local'] });

// set a cookie with options.
basil.set('def', 'uvw', { 'storages': ['cookie'], cookieSettings: { expireDays: 10, domain: 'mydomain.com', secure: true } });

// set 'xyz' value under 'abc' key in memory
basil.set('abc', 'xyz', { 'storages': ['memory'] });

// set value without JSON encoding
basil.set('foo', '{ "bar": "baz" }', { raw: true }); // will save { "bar": "baz" } as string

// retrieve keys
basil.keys(); // returns ['foo', 'abc']
basil.keys({ 'storages': ['memory'] }); // returns ['abc']

// retrieve keys map
basil.keysMap(); // returns { 'foo': ['local', 'cookie'], 'abc': ['memory'] }
basil.keysMap({ 'storages': ['memory'] }); // returns { 'abc': ['memory'] }
```

### Native storages
```
// Access native storages
// With basil API, but without namespace nor JSON parsing for values

// cookies
Basil.cookie.get(key);
Basil.cookie.set(key, value, { 'expireDays': days, 'domain': 'mydomain.com', 'secure': true });

// localStorage
Basil.localStorage.get(key);
Basil.localStorage.set(key, value);

// sessionStorage
Basil.sessionStorage.get(key);
Basil.sessionStorage.set(key, value);
```

### Namespaces

```javascript
basil = new window.Basil(options);

// store data under default namespace
basil.set('hello', 'world');

// store data under a given namespace
basil.set('hello', 42, { 'namespace': 'alt' });
basil.set('abc', 'def', { 'namespace': 'alt', 'storages': ['memory'] });

// retrieve data
basil.get('hello'); // return 'world'
basil.get('hello', { 'namespace': 'alt' }); // return 42

// retrieves keys
basil.keys(); // returns ['hello']
basil.keys({ 'namespace': 'alt' }); // returns ['hello', 'abc']

// retrieves  keys map
basil.keysMap(); // returns { 'hello': ['local'] }
basil.keysMap({ 'namespace': 'alt' }); // returns { 'hello': ['local'], 'abc': ['memory'] }

// remove data under a given namespace
basil.remove('hello', { 'namespace': 'alt' });
basil.get('hello'); // return 'world'
basil.get('hello', { 'namespace': 'alt' }); // return null

// reset data under a given namespace
basil.reset({ 'namespace': 'alt', 'storages': ['local', 'memory']});
```

## Configuration

Here is the whole `options` object that you could give to Basil:

```javascript
options = {
  // Namespace. Namespace your Basil stored data
  // default: 'b45i1'
  namespace: 'foo',

  // storages. Specify all Basil supported storages and priority order
  // default: `['local', 'cookie', 'session', 'memory']`
  storages: ['cookie', 'local']

  // expireDays. Default number of days before cookies expiration
  // default: 365
  expireDays: 31

  // keyDelimiter. The value used delimt the namespace from the key name
  // default: '.'
  keyDelimiter: '.'

};
```

## Compatibility

- Firefox 3.5+
- Internet Explorer 7 (requires [json2.js](//cdnjs.cloudflare.com/ajax/libs/json2/20130526/json2.min.js]))
- Internet Explorer 8+
- Chrome 4+
- Safari 4+


## Plugins

### List plugin

This plugin mimics [Redis Lists](http://redis.io/commands#list) methods and
behaviors. Here are the (yet) supported methods.

```javascript
basil = new window.Basil(options);
basil.lindex(key, index);
basil.linsert(key, where, pivot, value);
basil.llen(key);
basil.lpop(key);
basil.lpush(key, value);
basil.lrange(key, start, stop);
basil.lrem(key, count, value);
basil.lset(key, index, value);
basil.ltrim(key, start, stop);
basil.rpop(key);
basil.rpush(key, value);
```

### Set plugin

This plugin mimics [Redis Sets](http://redis.io/commands#set) methods and
behaviors. Except sscan all the methods are implemented.

```javascript
basil = new window.Basil(options);
basil.sadd(key, member [members ...]);
basil.scard(key);
basil.sdiff(key [keys ...]);
basil.sdiffstore(destination, key [keys ...]);
basil.sinter(key [keys ...]);
basil.sinterstore(destination, key [keys ...]);
basil.sismember(key, member);
basil.smember(key);
basil.smove(source, destination, member);
basil.spop(key);
basil.srandmember(key, [count]);
basil.srem(key, member [members ...]);
basil.sunion(key [keys ...]);
basil.sunionstore(destination, key [keys ...]);
```

## Build

To generate the production files, make sure you have already installed the
dependencies using ````npm install```` and then just use:

````
npm run-script build
````

## Tests

To launch the test suite, make sure you have already installed the dependencies
using ````npm install````.
Tests are launching in all your installed browsers. They're also launched on
Travis CI, in PhantomJS.

````
npm test
npm run-script test-plugins
````

## License

MIT. See `LICENSE.md`


CHANGELOG.md
# Basil changelog

## 0.4.5

  - Add options.keyDelimiter support

## 0.4.4

  - check third party cookies if not inside a top window
  - support string, number, boolean or array as key

## 0.4.3

  - [BC Break] `_toStoredKey()` is now using '.' separator instead of ':' (#35)

## 0.4.2

  - [BC Break] Encode and decode cookie value on set and get (#30, #35)

## 0.4.1

  - Added HTTPS cookie option (#27)

## 0.4.0

  - Updated README
  - Added `keys()` and `keysMap()` methods (#24)
  - fixed cookie/localStorage security error if cookies disabled
  - [BC Break] Basil raw storage engines do not need now a basil instance
    to be accessed. Use `Basil.cookie.set('foo', 'bar')` now instead of
    `new Basil().cookie.set('foo', 'bar')`


test/set.js
(function () {

	describe('Basil Set test suite', function () {
		var basil = new window.Basil();

		it('should have Basil Set plugin loaded', function () {
			expect(basil._setPlugin).to.be(true);
		});

		it('should have sadd method', function () {
			basil.sadd('basil_set', ['foo', 'bar']);
			expect(basil.smembers('basil_set')).to.eql(['foo', 'bar']);
			var addItem = basil.sadd('basil_set', 'baz');
			expect(addItem).to.be(1);
			expect(basil.smembers('basil_set')).to.eql(['foo', 'bar', 'baz']);
		});

		it('should not add duplicate elements', function () {
			basil.sadd('basil_set', ['foo', 'bar']);
			basil.sadd('basil_set', ['bar', 'baz']);
			expect(basil.smembers('basil_set')).to.eql(['foo', 'bar', 'baz']);
		});

		it('should have scard method', function() {
    		expect(basil.scard('basil_set')).to.be(0);
    		basil.sadd('basil_set', ['foo', 'bar']);
    		expect(basil.scard('basil_set')).to.be(2);
		});

		it('should have sdiff method', function() {
    		var array1 = ['a', 'b', 'c', 'd'], array2 = ['c'],
    			array3 = ['a', 'c', 'e'];

    		basil.sadd('basil_set_1', array1);
    		basil.sadd('basil_set_2', array2);
    		basil.sadd('basil_set_3', array3);

    		expect(basil.sdiff('basil_set_1', 'basil_set_2', 'basil_set_3')).to.eql(['b', 'd']);

    		basil.remove('basil_set_1');
    		basil.remove('basil_set_2');
    		basil.remove('basil_set_3');
		});

		it('should have sdiffstore method', function() {
			var array1 = ['a', 'b', 'c', 'd'], array2 = ['c'],
				array3 = ['a', 'c', 'e'];

			basil.sadd('basil_set_1', array1);
			basil.sadd('basil_set_2', array2);
			basil.sadd('basil_set_3', array3);

			var result = basil.sdiffstore('basil_diff_set', 'basil_set_1', 'basil_set_2', 'basil_set_3');
			expect(result).to.be(2);
			expect(basil.smembers('basil_diff_set')).to.eql(['b', 'd']);

			basil.remove('basil_diff_set');
			basil.remove('basil_set_1');
			basil.remove('basil_set_2');
			basil.remove('basil_set_3');
		});

		it('should have sinter method', function() {
    		var array1 = ['a', 'b', 'c', 'd'], array2 = ['c'],
				array3 = ['a', 'c', 'e'];

    		basil.sadd('basil_set_1', array1);
    		basil.sadd('basil_set_2', array2);
    		basil.sadd('basil_set_3', array3);

    		expect(basil.sinter('basil_set_1', 'basil_set_2', 'basil_set_3')).to.eql(['c']);

    		basil.remove('basil_set_1');
    		basil.remove('basil_set_2');
    		basil.remove('basil_set_3');
		});

		it('should have sinterstore method', function() {
    		var array1 = ['a', 'b', 'c', 'd'], array2 = ['c'],
				array3 = ['a', 'c', 'e'];

    		basil.sadd('basil_set_1', array1);
    		basil.sadd('basil_set_2', array2);
    		basil.sadd('basil_set_3', array3);

    		var result = basil.sinterstore('basil_intersect_set', 'basil_set_1', 'basil_set_2', 'basil_set_3');
    		expect(result).to.be(1);
    		expect(basil.smembers('basil_intersect_set')).to.eql(['c']);

    		basil.remove('basil_intersect_set');
    		basil.remove('basil_set_1');
    		basil.remove('basil_set_2');
    		basil.remove('basil_set_3');
		});

		it('should have sismember menthod', function() {
			basil.sadd('basil_set', ['foo', 'bar']);
			expect(basil.sismember('basil_set', 'foo')).to.be(1);
			expect(basil.sismember('basil_set', 'baz')).to.be(0);
		});

		it('should have smember menthod', function() {
			basil.sadd('basil_set', ['foo', 'bar']);
			expect(basil.smembers('basil_set')).to.eql(['foo', 'bar']);
			expect(basil.smembers('basil_set_2')).to.eql([]);
		});

		it('should have smove menthod', function() {
			basil.sadd('basil_set', ['foo', 'bar']);
			var result = basil.smove('basil_set', 'basil_set_2', 'bar');

			expect(result).to.be(1);
			expect(basil.smembers('basil_set')).to.eql(['foo']);
			expect(basil.smembers('basil_set_2')).to.eql(['bar']);

			result = basil.smove('basil_set', 'basil_set_2', 'foo');

			expect(result).to.be(1);
			expect(basil.smembers('basil_set')).to.eql([]);
			expect(basil.smembers('basil_set_2')).to.eql(['bar', 'foo']);

			result = basil.smove('basil_set', 'basil_set_2', 'foo');

			expect(result).to.be(0);

			basil.remove('basil_set');
			basil.remove('basil_set_2')
		});

		it('should have spop method', function() {
			var array = ['foo', 'bar', 'baz'];
			basil.sadd('basil_set', array);
			var result = basil.spop('basil_set');
			expect(array).to.contain(result);
			expect(basil.smembers('basil_set').length).to.be(2);
			basil.spop('basil_set');
			basil.spop('basil_set');
			result = basil.spop('basil_set');
			expect(result).to.be(null);
		});

		it('should have srandmember method', function() {
			var array = ['foo', 'bar', 'baz', 'qux', 'quux'];
			basil.sadd('basil_set', array);
			var result = basil.srandmember('basil_set');
			expect(array).to.contain(result);
			expect(basil.smembers('basil_set').length).to.be(5);

			result = basil.srandmember('basil_set', 3);
			expect(result).to.be.an(Array);
			expect(result.length).to.be(3);

			result = basil.srandmember('basil_set', 8);
			expect(result).to.be.an(Array);
			expect(result.length).to.be(5);

			result = basil.srandmember('basil_set', -8);
			expect(result).to.be.an(Array);
			expect(result.length).to.be(8);
		});

		it('should have srem method', function() {
			var array = ['foo', 'bar', 'baz', 'qux', 'quux'];
			basil.sadd('basil_set', array);

			var result = basil.srem('basil_set', 'baz');
			expect(result).to.be(1);
			expect(basil.smembers('basil_set')).to.eql(['foo', 'bar', 'qux', 'quux']);

			result = basil.srem('basil_set', ['baz', 'foo', 'quux']);
			expect(result).to.be(2);
			expect(basil.smembers('basil_set')).to.eql(['bar', 'qux']);

			basil.remove('basil_set');
			result = basil.srem('basil_set', 'baz');
			expect(result).to.be(0);
		});

		it('should have sunion method', function() {
			var array1 = ['a', 'b', 'c', 'd'], array2 = ['c'],
				array3 = ['a', 'c', 'e'];

			basil.sadd('basil_set_1', array1);
			basil.sadd('basil_set_2', array2);
			basil.sadd('basil_set_3', array3);

			expect(basil.sunion('basil_set_1', 'basil_set_2', 'basil_set_3')).to.eql(['a', 'b', 'c', 'd', 'e']);

			basil.remove('basil_set_1');
			basil.remove('basil_set_2');
			basil.remove('basil_set_3');
		});

		it('should have sunionstore method', function() {
			var array1 = ['a', 'b', 'c', 'd'], array2 = ['c'],
				array3 = ['a', 'c', 'e'];

			basil.sadd('basil_set_1', array1);
			basil.sadd('basil_set_2', array2);
			basil.sadd('basil_set_3', array3);

			var result = basil.sunionstore('basil_union_set', 'basil_set_1', 'basil_set_2', 'basil_set_3');
			expect(result).to.be(5);
			expect(basil.smembers('basil_union_set')).to.eql(['a', 'b', 'c', 'd', 'e']);

			basil.remove('basil_union_set');
			basil.remove('basil_set_1');
			basil.remove('basil_set_2');
			basil.remove('basil_set_3');
		});


        afterEach(function () {
			basil.remove('basil_set');
		});
	});

}());


test/list.js
(function () {

	describe('Basil List test suite', function () {
		var basil = new window.Basil();

		beforeEach(function () {
			basil.set('key', ['foo', 'bar']);
		});
		it('should have Basil List plugin loaded', function () {
			expect(basil._listPlugin).to.be(true);
		});
		it.skip('should have blpop method');
		it.skip('should have brpop method');
		it.skip('should have brpoplpush method');
		it('should have lindex method', function () {
			expect(basil.lindex('key', -42)).to.be(null);
			expect(basil.lindex('key', -1)).to.be('bar');
			expect(basil.lindex('key', 0)).to.be('foo');
			expect(basil.lindex('key', 1)).to.be('bar');
			expect(basil.lindex('key', 42)).to.be(null);
		});
		it('should have linsert method', function () {
			try {
				basil.linsert('key', 'AFTER');
				expect().fail("should not pass");
			} catch (e) {
				expect(e.message).to.be('ERR wrong number of arguments for \'linsert\' command');
			}

			try {
				basil.linsert('key', 'bar', 'foo', 'bar');
				expect().fail("should not pass");
			} catch (e) {
				expect(e.message).to.be('ERR syntax error');
			}

			try {
				basil.linsert('key', 'bar', { hello: 'world' }, 'bar');
				expect().fail("should not pass");
			} catch (e) {
				expect(e.message).to.be('ERR syntax error');
			}

			expect(basil.linsert('key', 'AFTER', 'foo', 'baz')).to.be(3);
			expect(basil.get('key')).to.be.eql(['foo', 'baz', 'bar']);
			expect(basil.linsert('key', 'BEFORE', 'foo', 'bux')).to.be(4);
			expect(basil.get('key')).to.be.eql(['bux', 'foo', 'baz', 'bar']);
			expect(basil.linsert('key', 'AFTER', 'donotexist', 'val')).to.be(-1);
			expect(basil.linsert('donotexist', 'AFTER', 'donotexist', 'val')).to.be(0);
		});
		it('should have llen method', function () {
			expect(basil.llen('key')).to.be(2);
			basil.set('key', []);
			expect(basil.llen('key')).to.be(0);
		});
		it('should have lpop method', function () {
			expect(basil.lpop('key')).to.be('foo');
			expect(basil.llen('key')).to.be(1);
			expect(basil.lpop('key')).to.be('bar');
			expect(basil.get('key')).to.be(null); // verify that key is well deleted if empty
		});
		it('should have lpush method', function () {
			expect(basil.lpush('key', 'baz')).to.be(3);
			expect(basil.get('key')).to.eql(['baz', 'foo', 'bar']);
		});
		it.skip('should have lpushx method');
		it('should have lrange method', function () {
			basil.rpush('key', 'baz');
			expect(basil.lrange('key', 0, 0)).to.eql(['foo']);
			expect(basil.lrange('key', 1, 1)).to.eql(['bar']);
			expect(basil.lrange('key', 2, 2)).to.eql(['baz']);
			expect(basil.lrange('key', 0, 2)).to.eql(['foo', 'bar', 'baz']);
			expect(basil.lrange('key', 0, 100)).to.eql(['foo', 'bar', 'baz']);
			expect(basil.lrange('key', -100, 100)).to.eql(['foo', 'bar', 'baz']);
			expect(basil.lrange('key', 5, 10)).to.eql([]);
		});
		it('should have lrem method', function () {
			try {
				basil.lrem('donotexist');
				expect().fail();
			} catch (e) {
				expect(e.message).to.be('ERR wrong number of arguments for \'lrem\' command');
			}

			try {
				basil.lrem('key', -2, { hello: 'world' });
				expect().fail();
			} catch (e) {
				expect(e.message).to.be('ERR syntax error');
			}

			expect(basil.lrem('donotexist', 2, 'foo')).to.be(0);

			var set = ['foo', 'foo', 'bar', 'baz', 'bar', 'foo', 'foo'];
			basil.set('key', set);
			expect(basil.lrem('key', -2, 'foo')).to.be(2);
			expect(basil.get('key')).to.eql(['foo', 'foo', 'bar', 'baz', 'bar']);

			basil.set('key', set);
			expect(basil.lrem('key', 1, 'bar')).to.be(1);
			expect(basil.get('key')).to.eql(['foo', 'foo', 'baz', 'bar', 'foo', 'foo']);

			basil.set('key', set);
			expect(basil.lrem('key', 0, 'foo')).to.be(4);
			expect(basil.get('key')).to.eql(['bar', 'baz', 'bar']);

			basil.set('key', set);
			expect(basil.lrem('key', -10, 'bar')).to.be(2);

		});
		it('should have lset method', function () {
			try {
				basil.lset('donotexist', 0, 'foo');
				expect.tail("should not pass");
			} catch (e) {
				expect(e.message).to.be('ERR no such key');
			}

			try {
				basil.lset('key', 2, 'baz');
				expect().fail("should not pass");
			} catch (e) {
				expect(e.message).to.be('ERR index out of range');
			}
			expect(basil.lset('key', 1, 'baz')).to.be('OK');
			expect(basil.lset('key', -2, 'bux')).to.be('OK');
			expect(basil.lrange('key', 0, 2)).to.eql(['bux', 'baz']);
		});
		it.skip('should have ltrim method');
		it('should have rpop method', function () {
			expect(basil.rpop('key')).to.be('bar');
			expect(basil.llen('key')).to.be(1);
			expect(basil.rpop('key')).to.be('foo');
			expect(basil.llen('key')).to.be(0);
			expect(basil.get('key')).to.be(null); // verify that key is well deleted if empty
		});
		it.skip('should have brpoplpush method');
		it('should have rpush method', function () {
			expect(basil.rpush('key', 'baz')).to.be(3);
			expect(basil.lindex('key', 2)).to.be('baz');
		});
		it.skip('should have rpushx method');
		afterEach(function () {
			basil.set('key', []);
		});
	});

}());


gulpfile.js
var gulp = require('gulp'),
    uglify = require('gulp-uglify'),
    rename = require('gulp-rename'),
    replace = require('gulp-replace');

gulp.task('default', function () {
    gulp.start('scripts');
});

gulp.task('scripts', ['default_scripts', 'es2015_src', 'es2015_min']);

gulp.task('default_scripts', function () {
    return gulp.src('src/basil.js')
        .pipe(gulp.dest('build'))
        .pipe(rename({ suffix: '.min' }))
        .pipe(uglify())
        .pipe(gulp.dest('build'));
});

gulp.task('es2015_src', function () {
    return gulp.src('src/basil.js')
        .pipe(rename({ suffix: '.es2015' }))
        .pipe(replace(/^(.|\n|\r)*?{\n/, ''))
        .pipe(replace(/window\.Basil =(.|\n|\r)*$/, 'export default Basil;'))
        .pipe(gulp.dest('build'));
});

gulp.task('es2015_min', function () {
    return gulp.src('src/basil.js')
        .pipe(rename({ suffix: '.es2015.min' }))
        .pipe(uglify()) // Uglify does not recognize ES2015 syntax.
        .pipe(replace(/^(.|\n|\r)*?var/, 'var'))
        .pipe(replace(/,window\.Basil=/, ';export default '))
        .pipe(replace(/(export default .)(.|\n|\r)*$/, '$1;'))
        .pipe(gulp.dest('build'));
});


src/basil.js
(function () {
	// Basil
	var Basil = function (options) {
		return Basil.utils.extend({}, Basil.plugins, new Basil.Storage().init(options));
	};

	// Version
	Basil.version = '0.4.5';

	// Utils
	Basil.utils = {
		extend: function () {
			var destination = typeof arguments[0] === 'object' ? arguments[0] : {};
			for (var i = 1; i < arguments.length; i++) {
				if (arguments[i] && typeof arguments[i] === 'object')
					for (var property in arguments[i])
						destination[property] = arguments[i][property];
			}
			return destination;
		},
		each: function (obj, fnIterator, context) {
			if (this.isArray(obj)) {
				for (var i = 0; i < obj.length; i++)
					if (fnIterator.call(context, obj[i], i) === false) return;
			} else if (obj) {
				for (var key in obj)
					if (fnIterator.call(context, obj[key], key) === false) return;
			}
		},
		tryEach: function (obj, fnIterator, fnError, context) {
			this.each(obj, function (value, key) {
				try {
					return fnIterator.call(context, value, key);
				} catch (error) {
					if (this.isFunction(fnError)) {
						try {
							fnError.call(context, value, key, error);
						} catch (error) {}
					}
				}
			}, this);
		},
		registerPlugin: function (methods) {
			Basil.plugins = this.extend(methods, Basil.plugins);
		},
		getTypeOf: function (obj) {
			if (typeof obj === 'undefined' || obj === null)
				return '' + obj;
			return Object.prototype.toString.call(obj).replace(/^\[object\s(.*)\]$/, function ($0, $1) { return $1.toLowerCase(); });
		}
	};
  	// Add some isType methods: isArguments, isBoolean, isFunction, isString, isArray, isNumber, isDate, isRegExp, isUndefined, isNull.
	var types = ['Arguments', 'Boolean', 'Function', 'String', 'Array', 'Number', 'Date', 'RegExp', 'Undefined', 'Null'];
	for (var i = 0; i < types.length; i++) {
		Basil.utils['is' + types[i]] = (function (type) {
			return function (obj) {
				return Basil.utils.getTypeOf(obj) === type.toLowerCase();
			};
		})(types[i]);
	}

	// Plugins
	Basil.plugins = {};

	// Options
	Basil.options = Basil.utils.extend({
		namespace: 'b45i1',
		storages: ['local', 'cookie', 'session', 'memory'],
		expireDays: 365,
		keyDelimiter: '.'
	}, window.Basil ? window.Basil.options : {});

	// Storage
	Basil.Storage = function () {
		var _salt = 'b45i1' + (Math.random() + 1)
				.toString(36)
				.substring(7),
			_storages = {},
			_isValidKey = function (key) {
				var type = Basil.utils.getTypeOf(key);
				return (type === 'string' && key) || type === 'number' || type === 'boolean';
			},
			_toStoragesArray = function (storages) {
				if (Basil.utils.isArray(storages))
					return storages;
				return Basil.utils.isString(storages) ? [storages] : [];
			},
			_toStoredKey = function (namespace, path, delimiter) {
				var key = '';
				if (_isValidKey(path)) {
					key += path;
				} else if (Basil.utils.isArray(path)) {
					path = Basil.utils.isFunction(path.filter) ? path.filter(_isValidKey) : path;
					key = path.join(delimiter);
				}
				return key && _isValidKey(namespace) ? namespace + delimiter + key : key;
 			},
			_toKeyName = function (namespace, key, delimiter) {
				if (!_isValidKey(namespace))
					return key;
				return key.replace(new RegExp('^' + namespace + delimiter), '');
			},
			_toStoredValue = function (value) {
				return JSON.stringify(value);
			},
			_fromStoredValue = function (value) {
				return value ? JSON.parse(value) : null;
			};

		// HTML5 web storage interface
		var webStorageInterface = {
			engine: null,
			check: function () {
				try {
					window[this.engine].setItem(_salt, true);
					window[this.engine].removeItem(_salt);
				} catch (e) {
					return false;
				}
				return true;
			},
			set: function (key, value, options) {
				if (!key)
					throw Error('invalid key');
				window[this.engine].setItem(key, value);
			},
			get: function (key) {
				return window[this.engine].getItem(key);
			},
			remove: function (key) {
				window[this.engine].removeItem(key);
			},
			reset: function (namespace) {
				for (var i = 0, key; i < window[this.engine].length; i++) {
					key = window[this.engine].key(i);
					if (!namespace || key.indexOf(namespace) === 0) {
						this.remove(key);
						i--;
					}
				}
			},
			keys: function (namespace, delimiter) {
				var keys = [];
				for (var i = 0, key; i < window[this.engine].length; i++) {
					key = window[this.engine].key(i);
					if (!namespace || key.indexOf(namespace) === 0)
						keys.push(_toKeyName(namespace, key, delimiter));
				}
				return keys;
			}
		};

		// local storage
		_storages.local = Basil.utils.extend({}, webStorageInterface, {
			engine: 'localStorage'
		});
		// session storage
		_storages.session = Basil.utils.extend({}, webStorageInterface, {
			engine: 'sessionStorage'
		});

		// memory storage
		_storages.memory = {
			_hash: {},
			check: function () {
				return true;
			},
			set: function (key, value, options) {
				if (!key)
					throw Error('invalid key');
				this._hash[key] = value;
			},
			get: function (key) {
				return this._hash[key] || null;
			},
			remove: function (key) {
				delete this._hash[key];
			},
			reset: function (namespace) {
				for (var key in this._hash) {
					if (!namespace || key.indexOf(namespace) === 0)
						this.remove(key);
				}
			},
			keys: function (namespace, delimiter) {
				var keys = [];
				for (var key in this._hash)
					if (!namespace || key.indexOf(namespace) === 0)
						keys.push(_toKeyName(namespace, key, delimiter));
				return keys;
			}
		};

		// cookie storage
		_storages.cookie = {
			check: function () {
				if (!navigator.cookieEnabled)
					return false;
				if (window.self !== window.top) {
					// we need to check third-party cookies;
					var cookie = 'thirdparty.check=' + Math.round(Math.random() * 1000);
					document.cookie = cookie + '; path=/';
					return document.cookie.indexOf(cookie) !== -1;
				}
				return true;
			},
			set: function (key, value, options) {
				if (!this.check())
					throw Error('cookies are disabled');
				options = options || {};
				options.cookieSettings = Basil.utils.extend({}, options.cookieSettings, options);
				if (!key)
					throw Error('invalid key');
				var cookie = encodeURIComponent(key) + '=' + encodeURIComponent(value);
				// handle expiration days
				if (options.cookieSettings.expireDays) {
					var date = new Date();
					date.setTime(date.getTime() + (options.cookieSettings.expireDays * 24 * 60 * 60 * 1000));
					cookie += '; expires=' + date.toGMTString();
				}
				// handle domain
				if (options.cookieSettings.domain && options.cookieSettings.domain !== document.domain) {
					var _domain = options.cookieSettings.domain.replace(/^\./, '');
					if (document.domain.indexOf(_domain) === -1 || _domain.split('.').length <= 1)
						throw Error('invalid domain');
					cookie += '; domain=' + options.cookieSettings.domain;
				}
				// handle secure
				if (options.cookieSettings.secure === true) {
					cookie += '; secure';
				}
				document.cookie = cookie + '; path=/';
			},
			get: function (key) {
				if (!this.check())
					throw Error('cookies are disabled');
				var encodedKey = encodeURIComponent(key);
				var cookies = document.cookie ? document.cookie.split(';') : [];
				// retrieve last updated cookie first
				for (var i = cookies.length - 1, cookie; i >= 0; i--) {
					cookie = cookies[i].replace(/^\s*/, '');
					if (cookie.indexOf(encodedKey + '=') === 0)
						return decodeURIComponent(cookie.substring(encodedKey.length + 1, cookie.length));
				}
				return null;
			},
			remove: function (key) {
				// remove cookie from main domain
				this.set(key, '', { expireDays: -1 });
				// remove cookie from upper domains
				var domainParts = document.domain.split('.');
				for (var i = domainParts.length; i >= 0; i--) {
					this.set(key, '', { expireDays: -1, domain: '.' + domainParts.slice(- i).join('.') });
				}
			},
			reset: function (namespace) {
				var cookies = document.cookie ? document.cookie.split(';') : [];
				for (var i = 0, cookie, key; i < cookies.length; i++) {
					cookie = cookies[i].replace(/^\s*/, '');
					key = cookie.substr(0, cookie.indexOf('='));
					if (!namespace || key.indexOf(namespace) === 0)
						this.remove(key);
				}
			},
			keys: function (namespace, delimiter) {
				if (!this.check())
					throw Error('cookies are disabled');
				var keys = [],
					cookies = document.cookie ? document.cookie.split(';') : [];
				for (var i = 0, cookie, key; i < cookies.length; i++) {
					cookie = cookies[i].replace(/^\s*/, '');
					key = decodeURIComponent(cookie.substr(0, cookie.indexOf('=')));
					if (!namespace || key.indexOf(namespace) === 0)
						keys.push(_toKeyName(namespace, key, delimiter));
				}
				return keys;
			}
		};

		return {
			init: function (options) {
				this.setOptions(options);
				return this;
			},
			setOptions: function (options) {
				this.options = Basil.utils.extend({}, this.options || Basil.options, options);
			},
			support: function (storage) {
				return _storages.hasOwnProperty(storage);
			},
			check: function (storage) {
				if (this.support(storage))
					return _storages[storage].check();
				return false;
			},
			set: function (key, value, options) {
				options = Basil.utils.extend({}, this.options, options);
				if (!(key = _toStoredKey(options.namespace, key, options.keyDelimiter)))
					return false;
				value = options.raw === true ? value : _toStoredValue(value);
				var where = null;
				// try to set key/value in first available storage
				Basil.utils.tryEach(_toStoragesArray(options.storages), function (storage, index) {
					_storages[storage].set(key, value, options);
					where = storage;
					return false; // break;
				}, null, this);
				if (!where) {
					// key has not been set anywhere
					return false;
				}
				// remove key from all other storages
				Basil.utils.tryEach(_toStoragesArray(options.storages), function (storage, index) {
					if (storage !== where)
						_storages[storage].remove(key);
				}, null, this);
				return true;
			},
			get: function (key, options) {
				options = Basil.utils.extend({}, this.options, options);
				if (!(key = _toStoredKey(options.namespace, key, options.keyDelimiter)))
					return null;
				var value = null;
				Basil.utils.tryEach(_toStoragesArray(options.storages), function (storage, index) {
					if (value !== null)
						return false; // break if a value has already been found.
					value = _storages[storage].get(key, options) || null;
					value = options.raw === true ? value : _fromStoredValue(value);
				}, function (storage, index, error) {
					value = null;
				}, this);
				return value;
			},
			remove: function (key, options) {
				options = Basil.utils.extend({}, this.options, options);
				if (!(key = _toStoredKey(options.namespace, key, options.keyDelimiter)))
					return;
				Basil.utils.tryEach(_toStoragesArray(options.storages), function (storage) {
					_storages[storage].remove(key);
				}, null, this);
			},
			reset: function (options) {
				options = Basil.utils.extend({}, this.options, options);
				Basil.utils.tryEach(_toStoragesArray(options.storages), function (storage) {
					_storages[storage].reset(options.namespace);
				}, null, this);
			},
			keys: function (options) {
				options = options || {};
				var keys = [];
				for (var key in this.keysMap(options))
					keys.push(key);
				return keys;
			},
			keysMap: function (options) {
				options = Basil.utils.extend({}, this.options, options);
				var map = {};
				Basil.utils.tryEach(_toStoragesArray(options.storages), function (storage) {
					Basil.utils.each(_storages[storage].keys(options.namespace, options.keyDelimiter), function (key) {
						map[key] = Basil.utils.isArray(map[key]) ? map[key] : [];
						map[key].push(storage);
					}, this);
				}, null, this);
				return map;
			}
		};
	};

	// Access to native storages, without namespace or basil value decoration
	Basil.memory = new Basil.Storage().init({ storages: 'memory', namespace: null, raw: true });
	Basil.cookie = new Basil.Storage().init({ storages: 'cookie', namespace: null, raw: true });
	Basil.localStorage = new Basil.Storage().init({ storages: 'local', namespace: null, raw: true });
	Basil.sessionStorage = new Basil.Storage().init({ storages: 'session', namespace: null, raw: true });


	// AMD export
	if (typeof define === 'function' && define.amd) {
		define(function() {
			return Basil;
		});
	// commonjs export
	} else if (typeof module !== 'undefined' && module.exports) {
		module.exports = Basil;
	// browser export
	} else {
		window.Basil = Basil;
	}

})();
